cwltool/workflow_job.py

```diff
@@ -879,3 +883,223 @@ def job(
+                                        MutableSequence[Optional[CWLOutputType]], []
+                                    )
+                                else:
+                                    self.output_buffer[iid] = None
+                    output_callback(self.output_buffer, self.processStatus)
+                    return
+                else:
+                    raise WorkflowException(
+                        "Loop condition 'loopWhen' must evaluate to 'true' or 'false'"
+                    )
+        except WorkflowException:
+            raise
+        except Exception:
+            _logger.exception("Unhandled exception")
+            self.processStatus = "permanentFail"
+            self.step.completed = True
+
+    def loop_callback(
+        self,
+        runtimeContext: RuntimeContext,
+        jobout: CWLObjectType,
+        processStatus: str,
+    ) -> None:
+        """Update the joborder object with output values from the last iteration."""
+        self.iteration += 1
+        try:
+            loop_req = cast(
+                CWLObjectType,
+                self.step.step.get_requirement("http://commonwl.org/cwltool#Loop")[0],
+            )
+            state: Dict[str, Optional[WorkflowStateItem]] = {}
+            for i in self.step.tool["outputs"]:
+                if "id" in i:
+                    iid = cast(str, i["id"])
+                    if iid in jobout:
+                        state[iid] = WorkflowStateItem(i, jobout[iid], processStatus)
+                        if loop_req.get("outputMethod") == "all":
+                            if iid not in self.output_buffer:
+                                self.output_buffer[iid] = cast(
+                                    MutableSequence[Optional[CWLOutputType]], []
+                                )
+                            cast(
+                                MutableSequence[Optional[CWLOutputType]],
+                                self.output_buffer[iid],
+                            ).append(jobout[iid])
+                        else:
+                            self.output_buffer[iid] = jobout[iid]
+                    else:
+                        _logger.error(
+                            "[%s] Output of iteration %i is missing expected field %s",
+                            self.step.name,
+                            self.iteration,
+                            iid,
+                        )
+                        processStatus = "permanentFail"
+            if _logger.isEnabledFor(logging.DEBUG):
+                _logger.debug(
+                    "Iteration %i of [%s] produced output %s",
+                    self.iteration,
+                    self.step.name,
+                    json_dumps(jobout, indent=4),
+                )
+
+            if processStatus not in ("success", "skipped"):
+                if self.processStatus != "permanentFail":
+                    self.processStatus = processStatus
```

Comment:
not quite following the logic here but if you follow my other comment you need to move this out so that `                    self.processStatus = processStatus` is always called